src/core/dyad_core.c

```diff
@@ -0,0 +1,622 @@
+    const char *prod_managed_path = ctx->prod_managed_path;
+    flux_kvs_txn_t *txn = NULL;
+    flux_future_t *f = NULL;
+    const size_t topic_len = PATH_MAX;
+    char topic[topic_len+1];
+    printf("Creating KVS key\n");
+    memset(topic, '\0', topic_len+1);
+    gen_path_key(
+            upath,
+            topic,
+            topic_len,
+            ctx->key_depth,
+            ctx->key_bins
+            );
+    if (ctx->h == NULL)
+    {
+        fprintf (stderr, "No Flux handle found in publish_via_flux!\n");
+        return DYAD_NOCTX;
+    }
+    // TODO FLUX_LOG_INFO
+    printf("Create KVS transaction\n");
+    txn = flux_kvs_txn_create();
+    if (txn == NULL)
+    {
+        DYAD_LOG_ERR(ctx, "Could not create Flux KVS transaction\n");
+        return DYAD_FLUXFAIL;
+    }
+    printf("Pack producer rank into transaction\n");
+    if (flux_kvs_txn_pack(txn, 0, topic, "i", ctx->rank) < 0)
+    {
+        DYAD_LOG_ERR(ctx, "Could not pack Flux KVS transaction\n");
+        return DYAD_FLUXFAIL;
+    }
+    printf("Calling dyad_kvs_commit\n");
+    int ret = dyad_kvs_commit(ctx, txn, upath, &f);
+    if (ret < 0)
+    {
+        DYAD_LOG_ERR(ctx, "dyad_kvs_commit failed!\n");
+        return ret;
+    }
+    return DYAD_OK;
+}
+
+int dyad_commit(dyad_ctx_t *ctx, const char *fname)
+{
+    int rc = validate_context(ctx, false, true);
+    if (DYAD_IS_ERROR(rc))
+    {
+        return rc;
+    }
+    if (ctx == NULL)
+    {
+        fprintf (stderr, "DYAD context not provided to dyad_commit\n");
+        return DYAD_NOCTX;
+    }
+    char upath[PATH_MAX] = {'\0'};
+    if (!cmp_canonical_path_prefix (ctx->prod_managed_path, fname, upath, PATH_MAX))
+    {
+        DYAD_LOG_INFO (ctx, "%s is not in the Producer's managed path\n", fname);
+        rc = 0;
+        goto commit_done;
+    }
+    ctx->reenter = false;
+    rc = publish_via_flux(ctx, upath);
+    ctx->reenter = true;
+
+commit_done:
+    if (rc == DYAD_OK && (ctx && ctx->check))
+    {
+        setenv(DYAD_CHECK_ENV, "ok", 1);
+    }
+    return rc;
+}
+
+#if DYAD_PERFFLOW
+__attribute__((annotate("@critical_path()")))
+#endif
+int dyad_kvs_lookup(dyad_ctx_t *ctx, const char* kvs_topic, uint32_t *owner_rank)
+{
+    int rc = -1;
+    flux_future_t *f = NULL;
+    f = flux_kvs_lookup(ctx->h,
+            ctx->kvs_namespace,
+            FLUX_KVS_WAITCREATE,
+            kvs_topic
+            );
+    if (f == NULL)
+    {
+        DYAD_LOG_ERR (ctx, "KVS lookup failed!\n");
+        return DYAD_BADLOOKUP;
+    }
+    rc = flux_kvs_lookup_get_unpack(f, "i", owner_rank);
+    if (f != NULL)
+    {
+        flux_future_destroy(f);
+        f = NULL;
+    }
+    if (rc < 0)
+    {
+        DYAD_LOG_ERR(ctx, "Could not unpack owner's rank from KVS response\n");
+        return DYAD_BADFETCH;
+    }
+    return DYAD_OK;
+}
+
+int dyad_fetch(dyad_ctx_t *ctx, const char* fname,
+        dyad_kvs_response_t **resp)
+{
+    int rc = validate_context(ctx, true, false);
```

Comment:
This is a redundant check. `dyad_consume()` already runs this check before calling this function.